File: sdk/php/src/Internal/Infra/DefaultWsService.php
<?php

namespace KuCoin\UniversalSDK\Internal\Infra;

use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Internal\Interfaces\Transport;
use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketClient;
use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketMessageCallback;
use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketService;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Internal\Utils\SubInfo;
use KuCoin\UniversalSDK\Model\ClientOption;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\WebSocketClientOption;
use KuCoin\UniversalSDK\Model\WebSocketEvent;
use KuCoin\UniversalSDK\Model\WsMessage;
use React\Promise\Deferred;
use React\Promise\PromiseInterface;
use RuntimeException;

class DefaultWsService implements WebSocketService
{
    /**
     * @var WebSocketClientOption $wsOption
     */
    private $wsOption;
    /**
     * @var bool $privateChannel
     */
    private $privateChannel;
    /**
     * @var Transport $tokenTransport
     */
    private $tokenTransport;
    /**
     * @var TopicManager $topicManager
     */
    private $topicManager;
    /**
     * @var WebSocketClient $client
     */
    private $client;
    /**
     * @var Serializer $serializer
     */
    private $serializer;

    public function __construct(ClientOption $option, $loop, $domainType, $privateChannel, $versionString)
    {
        if (!$option->websocketClientOption) {
            throw new RuntimeException("websocketClientOption is undefined");
        }

        $this->wsOption = $option->websocketClientOption;
        $this->privateChannel = $privateChannel;
        $this->tokenTransport = new DefaultTransport($option, $versionString);
        $this->topicManager = new TopicManager();
        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(
                    new JsonSerializedHandler()
                );
            })
            ->build();

        $this->client = new DefaultWebSocketClient(
            new DefaultWsTokenProvider($this->tokenTransport, $domainType, $privateChannel),
            $this->wsOption, $loop
        );

        $this->client->on("event", function (WebSocketEvent $event, $msg) {
            $this->emitEvent($event, $msg);
        });

        $this->client->on("message", function (WsMessage $message) {
            $this->processMessages($message);
        });

        $this->client->on("reconnected", function () {
            $this->recovery();
        });
    }

    private function emitEvent($event, $msg)
    {
        if (isset($this->wsOption->eventCallback)) {
            try {
                call_user_func($this->wsOption->eventCallback, $event, $msg);
            } catch (\Throwable $e) {
                error_log('event callback error: ' . $e->getMessage());
            }
        }
    }

    public function start(): PromiseInterface
    {
        return $this->client->start();
    }

    public function stop(): PromiseInterface
    {
        return $this->client->stop()->then(function () {
            $this->tokenTransport->close();
            $deferred = new Deferred();
            $deferred->resolve('');
            return $deferred->promise();
        });
    }

    public function subscribe(string $prefix, array $args, WebSocketMessageCallback $callback): PromiseInterface
    {
        $deferred = new Deferred();

        $subInfo = new SubInfo($prefix, $args, $callback);
        $subId = $subInfo->toId();
        $callbackManager = $this->topicManager->getCallbackManager($prefix);

        if (!$callbackManager->add($subInfo)) {
            $deferred->reject(new RuntimeException("Already subscribed: $subId"));
            return $deferred->promise();
        }

        $subEvent = new WsMessage();
        $subEvent->id = $subId;
        $subEvent->type = Constants::WS_MESSAGE_TYPE_SUBSCRIBE;
        $subEvent->topic = $subInfo->subTopic();
        $subEvent->privateChannel = $this->privateChannel;
        $subEvent->response = true;

        $this->client->write($subEvent, $this->wsOption->writeTimeout)->then(function () use ($deferred, $subId) {
            $deferred->resolve($subId);
        }, function ($err) use ($deferred, $callbackManager, $subId, $prefix) {
            $callbackManager->remove($subId);
            $deferred->reject($err);
        });

        return $deferred->promise();
    }

    public function unsubscribe($id): PromiseInterface
    {
        $deferred = new Deferred();

        $subInfo = SubInfo::fromId($id);
        $callbackManager = $this->topicManager->getCallbackManager($subInfo->prefix);

        $unsubEvent = new WsMessage();
        $unsubEvent->id = uniqid('', true);
        $unsubEvent->type = Constants::WS_MESSAGE_TYPE_UNSUBSCRIBE;
        $unsubEvent->topic = $subInfo->subTopic();
        $unsubEvent->privateChannel = $this->privateChannel;
        $unsubEvent->response = true;

        $this->client->write($unsubEvent, $this->wsOption->writeTimeout)->then(function () use ($callbackManager, $id, $deferred) {
            $callbackManager->remove($id);
            $deferred->resolve();
        }, function ($e) use ($deferred) {
            $deferred->reject($e);
        });

        return $deferred->promise();
    }

    private function processMessages(WsMessage $message)
    {
        $callbackManager = $this->topicManager->getCallbackManager($message->topic);

        $callback = $callbackManager->get($message->topic);
        if (!$callback) {
            return;
        }

        try {
            $callback->onMessage($message, $this->serializer);
        } catch (\Throwable $e) {
            $this->emitEvent(WebSocketEvent::EVENT_CALLBACK_ERROR, (string)$e);
        }
    }

    private function recovery()
    {
        $oldTopicManager = $this->topicManager;
        $this->topicManager = new TopicManager();

        $oldTopicManager->range(function ($key, $callbackManager) {
            foreach ($callbackManager->getSubInfo() as $sub) {
                if ($sub->callback) {
                    $this->subscribe($sub->prefix, $sub->args, $sub->callback)->then(function ($id) {
                        $this->emitEvent(WebSocketEvent::EVENT_RE_SUBSCRIBE_OK, $id);
                    }, function ($err) use ($sub) {
                        $this->emitEvent(WebSocketEvent::EVENT_RE_SUBSCRIBE_ERROR, (string)$err);
                    });
                }
            }
            return true;
        });
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/DefaultWsService.java
package com.kucoin.universal.sdk.internal.infra;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.internal.interfaces.*;
import com.kucoin.universal.sdk.model.*;
import java.util.Arrays;
import java.util.UUID;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public final class DefaultWsService implements WebSocketService, WebsocketTransportListener {
  private final ObjectMapper mapper = new ObjectMapper();
  private final WebsocketTransport client;
  private final WebSocketClientOption option;
  private final boolean privateChannel;
  private TopicManager topicManager = new TopicManager();

  public DefaultWsService(
      ClientOption opt, String domain, boolean privateChannel, String sdkVersion) {

    this.privateChannel = privateChannel;
    this.option = opt.getWebsocketClientOption();

    Transport tokenTransport = new DefaultTransport(opt, sdkVersion);

    WsTokenProvider tokenProvider =
        new DefaultWsTokenProvider(tokenTransport, domain, privateChannel);

    this.client = new DefaultWebsocketTransport(tokenProvider, option, this);
  }

  @Override
  public void start() {
    client.start();
  }

  @Override
  public void stop() {
    client.stop();
  }

  @Override
  public String subscribe(String prefix, String[] args, WebSocketMessageCallback<?> callback) {

    SubInfo sub = new SubInfo(prefix, Arrays.asList(args), callback);
    CallbackManager cm = topicManager.getCallbackManager(prefix);
    String id = sub.toId();
    Exception exception = null;

    if (!cm.add(sub)) {
      throw new IllegalStateException("already subscribed");
    }

    try {
      WsMessage msg = new WsMessage();
      msg.setId(id);
      msg.setType(Constants.WS_MESSAGE_TYPE_SUBSCRIBE);
      msg.setTopic(sub.subTopic());
      msg.setPrivateChannel(privateChannel);
      msg.setResponse(true);

      client.write(msg, option.getWriteTimeout()).join();
      return id;
    } catch (Exception e) {
      cm.remove(id);
      exception = e;
      throw new RuntimeException("subscribe failed", e);
    } finally {
      log.info(
          "subscribe prefix:{}, args:{}, private:{}, id:{}",
          prefix,
          args,
          privateChannel,
          id,
          exception);
    }
  }

  @Override
  public void unsubscribe(String id) {
    SubInfo sub = SubInfo.fromId(id);
    CallbackManager cm = topicManager.getCallbackManager(sub.subTopic());
    Exception exception = null;

    try {
      WsMessage msg = new WsMessage();
      msg.setId(UUID.randomUUID().toString());
      msg.setType(Constants.WS_MESSAGE_TYPE_UNSUBSCRIBE);
      msg.setTopic(sub.subTopic());
      msg.setPrivateChannel(privateChannel);
      msg.setResponse(true);

      client.write(msg, option.getWriteTimeout()).join();
      cm.remove(id);
    } catch (Exception e) {
      exception = e;
      throw new RuntimeException("unsubscribe failed", e);
    } finally {
      log.info("unsubscribe private:{}, id:{}", privateChannel, id, exception);
    }
  }

  @Override
  public void onEvent(WebSocketEvent event, String message) {
    notifyEvent(event, message);
  }

  @Override
  public void onMessage(WsMessage wsMessage) {
    CallbackManager cm = topicManager.getCallbackManager(wsMessage.getTopic());
    WebSocketMessageCallback<?> cb = cm.get(wsMessage.getTopic());
    if (cb == null) {
      log.warn("can not find callback manager, topic:{}", wsMessage.getTopic());
      return;
    }

    try {
      cb.onMessage(wsMessage, mapper);
    } catch (Throwable t) {
      notifyEvent(WebSocketEvent.CALLBACK_ERROR, t.getMessage());
    }
  }

  @Override
  public void onReconnected() {
    TopicManager oldTopicManager = topicManager;
    this.topicManager = new TopicManager();

    oldTopicManager.forEach(
        (key, value) -> {
          value
              .getSubInfo()
              .forEach(
                  sub -> {
                    try {
                      subscribe(
                          sub.getPrefix(),
                          sub.getArgs().toArray(new String[] {}),
                          sub.getCallback());
                      notifyEvent(WebSocketEvent.RE_SUBSCRIBE_OK, sub.toId());
                    } catch (Exception e) {
                      notifyEvent(WebSocketEvent.RE_SUBSCRIBE_ERROR, sub.toId());
                    }
                  });
        });
  }

  private void notifyEvent(WebSocketEvent ev, String msg) {
    if (option.getEventCallback() != null) {
      try {
        option.getEventCallback().onEvent(ev, msg);
      } catch (Exception e) {
        log.error("exception when notify event", e);
      }
    }
  }
}
